modules/bullet/rigid_body_bullet.cpp

```diff
@@ -998,15 +1017,14 @@ void RigidBodyBullet::reload_space_override_modificator() {
 
 	if (1 < countCombined) {
 		newGravity /= countCombined;
-		newLinearDamp /= countCombined;
-		newAngularDamp /= countCombined;
+		total_linear_damp /= countCombined;
+		total_angular_damp /= countCombined;
 	}
 
 	btVector3 newBtGravity;
 	G_TO_B(newGravity * gravity_scale, newBtGravity);
 
 	btBody->setGravity(newBtGravity);
-	btBody->setDamping(newLinearDamp, newAngularDamp);
```

Comment:
If you remove this line of code you are removing the damping set on the rigid body and the simulation will not take it into account it longer. 

The proper way would be ask in the bullet physics repository to add here https://github.com/bulletphysics/bullet3/blob/b25d806b1460b2d70d7603311d260ec5ae4363cd/src/BulletDynamics/Dynamics/btRigidBody.cpp#L144 a method that allows to customize the damping algorithm.

This is a useful link: https://code.google.com/archive/p/bullet/issues/74

Reply:
I'm not using the damping calculation inside of bullet. Instead I'm applying the changes on the velocity during the pre-tick callback (see `RigidBodyBullet::dispatch_callbacks`). It gives us full control over the damping algorithm so we can make sure it's consistent with Godot Physics, and we don't have to wait for changes made in the Bullet repository for fixing our bugs.

Applying the damping itself is just two lines of code, so I don't think it's a big deal to apply it in Godot codebase rather than using the algorithm in Bullet.

Reply:
Yes, but I still think this is not well handled there because the damping is still applied by the physics engine.

However, if you check the two link that I gave you is possible to see that the formula that you are using is already integrated into `applyDamping` in bullet.
This formula can be activated back by defining `USE_OLD_DAMPING_METHOD`, so you could define it and use that formula directly there.

However, it seems that the new formula added in the **2008** is the correct way of calculating the damping. I wonder if we should rely on the old one...

Reply:
Check this https://github.com/godotengine/godot/issues/19182#issuecomment-605417978

Reply:
Yes, https://code.google.com/archive/p/bullet/issues/74 is definitely the cause for the weird behavior.

I understand you would like to make it work from Bullet directly, but in this case I think it's going to be easier to solve it on Godot side.

I've checked the ticket you linked before I've made my changes, and as much as the new formula might be more physically accurate in some edge cases, it makes it much harder to use in the simple case developers face for most games, which is being able to setup values for damping on a linear scale, with fixed time steps in physics.
All the details are well explained in #19182: with the new formula, if you want to fine-tune the damping values, you would need some values extremely close to 1 and this is cumbersome to set in the editor.

As for using `USE_OLD_DAMPING_METHOD`, I've tried it too but it doesn't work out of the box because of the damping values being clamped to 0..1 (we need higher values than 1 to be able to set the damping correctly with this method):
https://github.com/bulletphysics/bullet3/blob/b25d806b1460b2d70d7603311d260ec5ae4363cd/src/BulletDynamics/Dynamics/btRigidBody.cpp#L137-L141

Reply:
Dump values higher than one breaks even your formula: `linear_vel *= MAX(0.0, 1.0 - delta_time * total_linear_damp);`

```
total_linear_damp = 1.5
delta_time = 1.0
1.0 - delta_time * total_linear_damp == -0.5
```
If you multiply the velocity by that value you invert the speed; so you need to rely on the function MAX to avoid this problem.

This is the reason why the damp can't be more than 1.

----

However, you may argue that `delta_time` is never 1; this doesn't matter since delta is not a value to scale the dumping, rather it's the time and the formula must work with any delta time span to be correct otherwise your formula will accumulate error overtime.

----

If you want to stop the body sooner you need a formula that scales the time, something like:
```
linear_vel *= (1.0 - delta_time * total_linear_damp) * damping_factor;
// OR
linear_vel *= 1.0 * time_scale - delta_time * time_scale * total_linear_damp;
```

Where `total_linear_damp` is always something between 0 and 1

Reply:
I'm applying the damping the exact same way it's done in Godot Physics:
https://github.com/godotengine/godot/blob/fcfffd729789cd80aa77056ca089697b52297d04/servers/physics_3d/body_3d_sw.cpp#L536-L547

It works in Godot Physics, and it works in my tests for this PR, with different values ranging from 0 to 100. As you mentioned, if the delta time was 1 the values would become negative, so they are clamped to 0 (full damping) by the MAX function. However, in most cases, delta time is a fixed value smaller than 1 so damping values over 1 make sense.

Reply:
It works with these settings in my own project :)
(note that it needs PR #37316 as well)
<img src="https://user-images.githubusercontent.com/1075032/77820845-a0740080-70e5-11ea-933b-8d2c58fbfc13.gif" width="200" />
<img src="https://user-images.githubusercontent.com/1075032/77820847-a4078780-70e5-11ea-95c5-0b88fcf5c421.png" width="200" />

Reply:
According to this https://github.com/godotengine/godot/issues/19182#issuecomment-605435823 implementing custom damping in the pre-tick callback is a valid solution, but we could also go for the #ifdef inside Bullet since we can to the change immediately and it will be merged to the Bullet repo.

Reply:
Yes, I still don't like to process the damping twice, however I want to find a solution for this problem.

So the ifdef solution is nice, and if Erwin is ok to add a custom equation into bullet even better.
However, what concern me is:
- I'm not a big fan of the proposed equation, since it is correct only under certain assumptions.
- Why the current Bullet damping is so weird? Is it correct and our expectations wrong or what?

So, I asked for more info about it to Erwin, we can decide how to proceed after we receive an answer (maybe that the damping doesn't work correctly due of another bug or that the new equation after all is not so correct)... So let's see.

Reply:
Sounds good to me!

Concerning the equation itself:
I understand your concern with the simplified version, but I would rather have one by default that's usable in most practical cases by godot users, and let them implement a custom damping equation in scripts in the rare cases where they would need more physically accurate results. We could document the limitations to make it clear.

Also, having the same behavior independently of the physics engine is better, so whatever we come up with should be probably used in both Bullet & Godot Physics to make it consistent.

Reply:
So we didn't receive an answer for `Why the current Bullet damping is so weird? Is it correct and our expectations wrong or what?` so I propose to move this issue forward and fix it.

I think that the damping should always go between 0 and 1 (similarly on the Friction and Bounciness). Damping 1 mean that the body is completely stopped in 1 second. If the user would really need to stop the body sooner, it's likely that the damping is not the right tool to use, and it may want to add external forces.

So I would like to clamp the damping, on editor and on the Rigid Body, and use the bullet **legacy** equation. By clamping the damping, also Godot Physics will behave like Bullet Physics without further changes.

Reply:
Thanks, I agree with moving on with the bullet legacy equation. It will require a few changes within bullet code to compile so I'll make a PR on the bullet repo as well.

About the 0..1 clamping though, I don't see a good practical reason to do it. After the changes in the equation, Bullet and Godot will be working the exact same way, so there's no need to make further changes that would only cause a large range of use cases not to be supported anymore.

Using damping to stop a rigid body from moving faster than 1 second is a pretty common use case for games. Two examples that immediately come to my mind are:
- Physics-based character controllers have to stop quickly in order to avoid input/animation lag
- Joint stabilization for secondary animation often requires aggressive damping (see my example above)

Reply:
Well, rethinking on what's going on you are correct about the fact that we remove a range of use cases by constrain the time to 1 sec.

I would like that it is properly integrated, by adding the time scale to the equation, but Erwin doesn't want to change the API.

So you are right about the fact that the formula `linear_vel *= MAX(0.0, 1.0 - delta_time * total_linear_damp);` even if not perfect is a good compromise.

So, ok let's use that equation.

Reply:
Cool! Thanks for taking the time to come to an agreement that's a good compromise for users.